cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -270,10 +274,26 @@ public boolean apply(RunRecordMeta record) {
   }
 
   @Nullable
-  private RuntimeInfo createRuntimeInfo(ProgramId programId, TwillController controller, RunId runId) {
+  private RuntimeInfo createRuntimeInfo(final ProgramId programId, TwillController controller, final RunId runId) {
     try {
       ProgramDescriptor programDescriptor = store.loadProgram(programId);
       ProgramController programController = createController(programDescriptor, controller, runId);
+      // Add a listener that publishes KILLED status notification when the YARN application is killed in case that
+      // the KILLED status notification is not published from the YARN application container, so we don't need to wait
+      // for the run record corrector to mark the status as KILLED.
+      programController.addListener(new AbstractListener() {
```

Comment:
also, this seems like it could result in a NPE.
The line below this `addListener` statement does a null check against `programController`.